### DISCUSSION TITLE
Applying distributions when generating collections

### DISCUSSION THREAD

**Discussion by davidkeaveny** [bchavez/Bogus#541] [Ideas]
Has anyone ever had a need to apply non-even distributions when selecting random values?

For instance, if I were creating a fake dataset of passengers arriving at the international terminal of London Heathrow airport, I might want it to be semi-realistic in that passengers held passports in the following distribution:

* 30% British
* 30% European
* 20% Asian
* 10% American
* 10% Other

This obviously would allow us to model e.g. passenger flow through immigration, with its different channels depending on the passport you hold.

If I were to define an enum, then I could use the `PickRandom<T>` method, but that would give an even distribution. 

```csharp
public enum PassportCategory
{
  British = 0,
  European = 1, 
  Asian= 2,
  American = 3,
  Other = 4
}

var faker = new Faker<Passenger>()
    .RuleFor(passenger => passenger.Category, fake => fake.PickRandom<PassportCategory>());
var passengers = faker.Generate(1000);
```

I could define a map that gives the distribution I want, and pick from that, but that doesn't scale well:

```csharp
var map = new int[] { 0, 0, 0, 1, 1, 1, 2, 2, 3, 4 };
var faker = new Faker<Passenger>()
    .RuleFor(passenger => passenger.Category, fake => (PassportCategory) fake.PickRandom(map));
```

**Comment by bchavez** [comment]
Hi @davidkeaveny , there is a `f => f.Random.WeightedRandom<T>(items[], weights[])` method for weighted randomness.

Here's an example:

```csharp
void Main()
{
   var f = new Faker();
   var items = Enum.GetValues<PassportCategory>();
   items.Dump();
   float[] weights = [0.30f, 0.30f, 0.20f, 0.10f, 0.10f];
   
   var selections = Enumerable.Range(1,100)
      .Select( _ => f.Random.WeightedRandom<PassportCategory>(items, weights))
      .ToArray();
      
   selections.Count().Dump();
      
   var british = selections.Count(i => i is PassportCategory.British);
   var euro = selections.Count(i => i is PassportCategory.European);
   var asian = selections.Count(i => i is PassportCategory.Asian);
   var american = selections.Count(i => i is PassportCategory.American);
   var other = selections.Count(i => i is PassportCategory.Other);
   
   british.Dump();
   euro.Dump();
   asian.Dump();
   american.Dump();
   other.Dump();
}
public enum PassportCategory
{
   British = 0,
   European = 1,
   Asian = 2,
   American = 3,
   Other = 4
}
```
```
OUTPUT:
British
European
Asian
American
Other

100

24 // british
34 // european
16 // asian
13 // american
13 // other
```
Of course, your numbers might not be exactly the same, but similar since I'm not setting a seed value. Hope that helps.

**Reply by davidkeaveny** [reply]
Thanks @bchavez that's very helpful!
